pandas/core/arrays/categorical.py

```diff
@@ -2494,6 +2494,20 @@ def isin(self, values):
         code_values = code_values[null_mask | (code_values >= 0)]
         return algorithms.isin(self.codes, code_values)
 
+    def replace(self, to_replace, value, inplace=False):
+        inplace = validate_bool_kwarg(inplace, "inplace")
+        cat = self if inplace else self.copy()
+        categories = cat.categories.tolist()
+        if to_replace in categories:
```

Comment:
what happens if `to_replace` is not hashable?  I think at the Categorical level we are OK raising, but at the Block level we would cast to object right?

Reply:
Would casting to object help?
```
>>> import pandas as pd
>>> s=pd.Series([1,2]).astype(object)
>>> s.replace(1,[1,2,3])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/core/series.py", line 4303, in replace
    method=method,
  File "pandas/core/generic.py", line 6819, in replace
    raise TypeError(msg)  # pragma: no cover
TypeError: Invalid "to_replace" type: 'int'
```

Reply:
@jreback I've merged master and fixed all comments except this one for the reason above